File: examples/ios/UniRTChatExample/UniRTChatExample/ChatViewModel.swift
// Copyright (c) 2026 Peter Huang.
// SPDX-License-Identifier: BSD-3-Clause

import CUniRT
import Foundation
import UniRTKit

enum ChatMode: String, CaseIterable {
    case text = "Text"
    case vision = "Vision"
}

struct DisplayMessage: Identifiable {
    let id = UUID()
    let role: String
    var text: String
    var imagePath: String?
}

@MainActor
final class ChatViewModel: ObservableObject {
    @Published var mode: ChatMode = .text
    @Published var messages: [DisplayMessage] = []
    @Published var input: String = ""
    @Published var status: String = "loading model..."
    @Published var isBusy: Bool = false
    @Published var modelName: String = ""
    @Published var stats: LlmRuntimeStats?
    @Published var attachedImagePath: String?

    private var llmSession: LlmSession?
    private var vlmSession: VlmSession?
    private var llmHistory: [ChatMessage] = []
    private var vlmHistory: [VlmChatMessage] = []
    private var llmMessages: [DisplayMessage] = []
    private var vlmMessages: [DisplayMessage] = []
    private var statsTask: Task<Void, Never>?
    private var pluginRegistered = false

    var testImagePath: String? { Bundle.main.path(forResource: "test-photo", ofType: "jpg") }

    func start() async {
        await switchMode(to: .text)
    }

    /// Both sessions, once opened, stay resident for the app's lifetime —
    /// switching modes never re-loads or frees a model. Loading the VLM's
    /// ~300MB of weights is enough of a momentary memory spike that iOS has
    /// been observed reclaiming memory from (and killing) the Control Center
    /// screen-recording broadcast process right as a switch completes;
    /// paying that cost once per mode instead of on every toggle makes that
    /// risk a one-time thing rather than a per-switch one. Warm up both
    /// modes once before recording a demo and every switch after that is
    /// instant with no new memory churn.
    func switchMode(to newMode: ChatMode) async {
        mode = newMode
        attachedImagePath = nil

        if newMode == .text, llmSession != nil {
            messages = llmMessages
            stats = try? await llmSession?.runtimeStats()
            status = "ready (\(UniRT.plugins.joined(separator: ", ")))"
            return
        }
        if newMode == .vision, let session = vlmSession {
            messages = vlmMessages
            stats = try? await session.runtimeStats()
            let caps = try? await session.capabilities()
            status = "ready (\(UniRT.plugins.joined(separator: ", "))) — vision: \(caps?.supportsVision == true)"
            return
        }

        statsTask?.cancel()
        messages = []
        stats = nil
        status = "loading model..."

        do {
            if !pluginRegistered {
                try UniRT.registerStaticPlugin(identity: unirt_plugin_id, open: unirt_plugin_open)
                try UniRT.start()
                pluginRegistered = true
            }
            switch newMode {
            case .text:
                guard let modelPath = Bundle.main.path(forResource: "model", ofType: "gguf") else {
                    status = "model.gguf not bundled — see examples/ios/UniRTChatExample/README.md"
                    return
                }
                modelName = (modelPath as NSString).lastPathComponent
                llmSession = try await UniRT.createLlmSession(modelPath: modelPath, nCtx: 2048)
                status = "ready (\(UniRT.plugins.joined(separator: ", ")))"
            case .vision:
                guard let modelPath = Bundle.main.path(forResource: "vlm-model", ofType: "gguf") else {
                    status = "vlm-model.gguf not bundled — see examples/ios/UniRTChatExample/README.md"
                    return
                }
                let mmprojPath = Bundle.main.path(forResource: "mmproj", ofType: "gguf")
                modelName = (modelPath as NSString).lastPathComponent
                vlmSession = try await UniRT.createVlmSession(modelPath: modelPath, mmprojPath: mmprojPath, nCtx: 2048)
                let caps = try await vlmSession?.capabilities()
                status = "ready (\(UniRT.plugins.joined(separator: ", "))) — vision: \(caps?.supportsVision == true)"
            }
            startStatsPolling()
        } catch {
            status = "load failed: \(error)"
        }
    }

    private func startStatsPolling() {
        statsTask?.cancel()
        statsTask = Task { [weak self] in
            while !Task.isCancelled {
                guard let self else { return }
                if !self.isBusy {
                    switch self.mode {
                    case .text:
                        self.stats = try? await self.llmSession?.runtimeStats()
                    case .vision:
                        self.stats = try? await self.vlmSession?.runtimeStats()
                    }
                }
                try? await Task.sleep(nanoseconds: 2_000_000_000)
            }
        }
    }

    func attachTestImage() {
        attachedImagePath = testImagePath
    }

    func clearAttachment() {
        attachedImagePath = nil
    }

    func send() {
        let text = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty, !isBusy else { return }
        input = ""
        isBusy = true
        status = "generating..."

        switch mode {
        case .text:
            sendText(text)
        case .vision:
            sendVision(text)
        }
    }

    private func sendText(_ text: String) {
        guard let session = llmSession else { isBusy = false; return }
        llmHistory.append(.user(text))
        messages.append(DisplayMessage(role: "user", text: text))
        let replyIndex = messages.count
        messages.append(DisplayMessage(role: "assistant", text: ""))

        Task {
            do {
                let prompt = try await session.applyChatTemplate(llmHistory)
                var full = ""
                for try await piece in session.stream(prompt: prompt, options: GenerateOptions(maxTokens: 128)) {
                    full += piece
                    messages[replyIndex].text = full
                }
                llmHistory.append(.assistant(full))
                status = "ready"
            } catch {
                messages[replyIndex].text = "error: \(error)"
                status = "ready"
            }
            isBusy = false
            llmMessages = messages
        }
    }

    private func sendVision(_ text: String) {
        guard let session = vlmSession else { isBusy = false; return }
        let imagePath = attachedImagePath
        attachedImagePath = nil

        let contents: [ContentPart] = imagePath.map { [.text(text), .image(path: $0)] } ?? [.text(text)]
        vlmHistory.append(VlmChatMessage(role: "user", contents: contents))
        messages.append(DisplayMessage(role: "user", text: text, imagePath: imagePath))
        let replyIndex = messages.count
        messages.append(DisplayMessage(role: "assistant", text: ""))

        Task {
            do {
                let prompt = try await session.applyChatTemplate(vlmHistory)
                let options = VlmGenerateOptions(maxTokens: 160, imagePaths: imagePath.map { [$0] } ?? [])
                var full = ""
                for try await piece in session.stream(prompt: prompt, options: options) {
                    full += piece
                    messages[replyIndex].text = full
                }
                vlmHistory.append(.assistant(full))
                status = "ready"
            } catch {
                messages[replyIndex].text = "error: \(error)"
                status = "ready"
            }
            isBusy = false
            vlmMessages = messages
        }
    }
}

func formatBytes(_ n: Int64?) -> String {
    guard let n, n >= 0 else { return "n/a" }
    if n == 0 { return "0 B" }
    let units = ["B", "KB", "MB", "GB"]
    var value = Double(n)
    var unitIndex = 0
    while value >= 1024, unitIndex < units.count - 1 {
        value /= 1024
        unitIndex += 1
    }
    let decimals = (value >= 10 || unitIndex == 0) ? 0 : 1
    return String(format: "%.\(decimals)f %@", value, units[unitIndex])
}

File: examples/android/UniRTChatExample/app/src/main/kotlin/ai/unirt/example/chat/ChatViewModel.kt
// Copyright (c) 2026 Peter Huang.
// SPDX-License-Identifier: BSD-3-Clause

package ai.unirt.example.chat

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import ai.unirt.ChatMessage
import ai.unirt.ContentPart
import ai.unirt.GenerateOptions
import ai.unirt.GenerationProfile
import ai.unirt.LlmSession
import ai.unirt.LlmStreamResult
import ai.unirt.RuntimeStats
import ai.unirt.UniRT
import ai.unirt.VlmChatMessage
import ai.unirt.VlmGenerateOptions
import ai.unirt.VlmSession
import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

enum class ChatMode(val label: String) { Text("Text"), Vision("Vision") }

data class DisplayMessage(val role: String, val text: String, val imagePath: String? = null)

/**
 * Mirrors the iOS example's ChatViewModel: two sessions (LLM + VLM), both
 * kept resident once loaded so switching modes never re-pays model load or
 * its memory spike; per-mode transcript and display history; streaming
 * replies appended token-by-token.
 *
 * Models ship in the APK's assets (see README) but the native layer wants
 * filesystem paths, so each asset is copied to filesDir once on first use.
 */
class ChatViewModel(application: Application) : AndroidViewModel(application) {
    var mode by mutableStateOf(ChatMode.Text)
        private set
    val messages = mutableStateListOf<DisplayMessage>()
    var input by mutableStateOf("")
    var status by mutableStateOf("loading model...")
        private set
    var isBusy by mutableStateOf(false)
        private set
    var modelName by mutableStateOf("")
        private set
    var profile by mutableStateOf<GenerationProfile?>(null)
        private set
    var stats by mutableStateOf<RuntimeStats?>(null)
        private set
    var attachedImagePath by mutableStateOf<String?>(null)
        private set

    private var llmSession: LlmSession? = null
    private var vlmSession: VlmSession? = null
    private val llmHistory = mutableListOf<ChatMessage>()
    private val vlmHistory = mutableListOf<VlmChatMessage>()
    private var llmMessages = listOf<DisplayMessage>()
    private var vlmMessages = listOf<DisplayMessage>()
    private var started = false
    private var statsJob: Job? = null

    val testImagePath: String? get() = ensureAsset("test-photo.jpg")

    fun start() {
        viewModelScope.launch { switchMode(ChatMode.Text) }
    }

    fun switchMode(newMode: ChatMode) {
        if (isBusy) return
        mode = newMode
        attachedImagePath = null

        llmSession?.takeIf { newMode == ChatMode.Text }?.let {
            messages.setAll(llmMessages)
            status = readyStatus()
            return
        }
        vlmSession?.takeIf { newMode == ChatMode.Vision }?.let {
            messages.setAll(vlmMessages)
            status = readyStatus()
            return
        }

        messages.clear()
        profile = null
        status = "loading model..."
        isBusy = true

        viewModelScope.launch {
            try {
                if (!started) {
                    UniRT.start()
                    started = true
                }
                when (newMode) {
                    ChatMode.Text -> {
                        val modelPath = withContext(Dispatchers.IO) { ensureAsset("model.gguf") }
                        if (modelPath == null) {
                            status = "model.gguf not bundled — see examples/android/UniRTChatExample/README.md"
                            return@launch
                        }
                        modelName = File(modelPath).name
                        llmSession = UniRT.createLlmSession(modelPath, nCtx = 2048)
                    }
                    ChatMode.Vision -> {
                        val modelPath = withContext(Dispatchers.IO) { ensureAsset("vlm-model.gguf") }
                        if (modelPath == null) {
                            status = "vlm-model.gguf not bundled — see examples/android/UniRTChatExample/README.md"
                            return@launch
                        }
                        modelName = File(modelPath).name
                        val mmprojPath = withContext(Dispatchers.IO) { ensureAsset("mmproj.gguf") }
                        vlmSession = UniRT.createVlmSession(modelPath, mmprojPath = mmprojPath, nCtx = 2048)
                    }
                }
                status = readyStatus()
                startStatsPolling()
            } catch (e: Exception) {
                status = "load failed: ${e.message}"
            } finally {
                isBusy = false
            }
        }
    }

    /** Live memory/device stats every 2 s while idle — generation owns the
     *  session's single native thread, so polling waits it out via isBusy. */
    private fun startStatsPolling() {
        statsJob?.cancel()
        statsJob = viewModelScope.launch {
            while (isActive) {
                if (!isBusy) {
                    stats = try {
                        when (mode) {
                            ChatMode.Text -> llmSession?.runtimeStats()
                            ChatMode.Vision -> vlmSession?.runtimeStats()
                        }
                    } catch (_: Exception) {
                        null
                    }
                }
                delay(2_000)
            }
        }
    }

    fun attachTestImage() {
        attachedImagePath = testImagePath
    }

    fun clearAttachment() {
        attachedImagePath = null
    }

    fun send() {
        val text = input.trim()
        if (text.isEmpty() || isBusy) return
        input = ""
        isBusy = true
        status = "generating..."
        when (mode) {
            ChatMode.Text -> sendText(text)
            ChatMode.Vision -> sendVision(text)
        }
    }

    private fun sendText(text: String) {
        val session = llmSession ?: run { isBusy = false; return }
        llmHistory += ChatMessage.user(text)
        messages += DisplayMessage("user", text)
        val replyIndex = messages.size
        messages += DisplayMessage("assistant", "")

        viewModelScope.launch {
            try {
                val prompt = session.applyChatTemplate(llmHistory)
                var full = ""
                session.stream(prompt, GenerateOptions(maxTokens = 128)).collect { event ->
                    when (event) {
                        is LlmStreamResult.Token -> {
                            full += event.text
                            messages[replyIndex] = messages[replyIndex].copy(text = full)
                        }
                        is LlmStreamResult.Completed -> profile = event.profile
                        is LlmStreamResult.Error -> throw event.cause
                    }
                }
                llmHistory += ChatMessage.assistant(full)
                status = readyStatus()
            } catch (e: Exception) {
                messages[replyIndex] = messages[replyIndex].copy(text = "error: ${e.message}")
                status = readyStatus()
            }
            isBusy = false
            llmMessages = messages.toList()
        }
    }

    private fun sendVision(text: String) {
        val session = vlmSession ?: run { isBusy = false; return }
        val imagePath = attachedImagePath
        attachedImagePath = null

        val contents = if (imagePath != null) {
            listOf(ContentPart.Text(text), ContentPart.Image(imagePath))
        } else {
            listOf(ContentPart.Text(text))
        }
        vlmHistory += VlmChatMessage("user", contents)
        messages += DisplayMessage("user", text, imagePath)
        val replyIndex = messages.size
        messages += DisplayMessage("assistant", "")

        viewModelScope.launch {
            try {
                val prompt = session.applyChatTemplate(vlmHistory)
                // The template re-renders the whole transcript, so every image
                // marker in it — past turns included — needs its file supplied
                // again, in order; sending only the new image fails with
                // "prompt has N media markers but M files were supplied".
                // The prefix cache keeps re-supplied old images cheap.
                val allImages = vlmHistory.flatMap { it.contents }
                    .filterIsInstance<ContentPart.Image>()
                    .map { it.path }
                val options = VlmGenerateOptions(maxTokens = 160, imagePaths = allImages)
                var full = ""
                session.stream(prompt, options).collect { event ->
                    when (event) {
                        is LlmStreamResult.Token -> {
                            full += event.text
                            messages[replyIndex] = messages[replyIndex].copy(text = full)
                        }
                        is LlmStreamResult.Completed -> profile = event.profile
                        is LlmStreamResult.Error -> throw event.cause
                    }
                }
                vlmHistory += VlmChatMessage.assistant(full)
                status = readyStatus()
            } catch (e: Exception) {
                messages[replyIndex] = messages[replyIndex].copy(text = "error: ${e.message}")
                status = readyStatus()
            }
            isBusy = false
            vlmMessages = messages.toList()
        }
    }

    private fun readyStatus() = "ready (${UniRT.plugins().joinToString(", ")})"

    /** Copy an APK asset to filesDir once; null when the asset isn't bundled. */
    private fun ensureAsset(name: String): String? {
        val context = getApplication<Application>()
        val target = File(context.filesDir, name)
        return try {
            context.assets.open(name).use { input ->
                val size = input.available().toLong()
                if (!target.exists() || target.length() != size) {
                    target.outputStream().use { input.copyTo(it) }
                }
            }
            target.absolutePath
        } catch (_: java.io.FileNotFoundException) {
            null
        }
    }

    private fun <T> MutableList<T>.setAll(items: List<T>) {
        clear()
        addAll(items)
    }
}
